### DISCUSSION TITLE
Hitting "Timed out fetching a new connection from the connection pool" error

### DISCUSSION THREAD

**Discussion by arnmishra** [prisma/prisma#22252] [Q&A]
### Question

I've read pretty much any prisma documentation, github issue, q&a discussion, stack overflow, etc on this but haven't been able to come up with what is causing our issue. 

We have been seeing the following error `Timed out fetching a new connection from the connection pool. More info: http://pris.ly/d/connection-pool (Current connection pool timeout: 10, connection limit: 5)`. This is happening on a process running in a google cloud run container. I understand that this means we are running out of available connections. We are using prisma client 5 on a postgres db in cloud sql.

We have a set of 5 cloud run instances that all speak to our postgres database. All 5 have their own prisma client that has a connection pool with a default limit of 5 (2 cores). The failures all have been happening so far in the same cloud run service that runs async jobs for us. This service runs a small typescript app that triggers jobs using BullMQ (https://docs.bullmq.io/). 

In the course of 17 attempts to trigger a job in the last 3 days: 
- 13 had no issues
- 3 failed but worked on 1 retry
- 1 failed on all 3 retries and only worked after a 4th manual try
(all failures were the same prisma client timeout)

Every time the query that fails is on the very first query made to the db in that container. At the time of the failure, there is no other activity in that container, and only a very minimal light load, if any, on the other 4 cloud run services (which all have their own prisma connection pool). Therefore, I can't imagine what would cause a timeout. I thought it may have to do with cold starts somehow but we've configured all our cloud run containers with a minimum of 1 instance always running.

### Information about Prisma Schema, Client Queries and Environment (optional)

```
datasource db {
  provider = "postgresql"
  url      = env("POSTGRES_URL")
}

generator client {
  provider        = "prisma-client-js"
}
```

The query its failing on is a simple `findUniqueOrThrow`

- Database: PostgresQL
- Node.js version: 18.16

- Run `prisma -v` to see your Prisma version and paste it
5.2.0

**Comment by arnmishra** [comment]
more info, I added a log of the metrics on the prisma client after the prisma query.

```
prismaClient.$on('query'), (e) => prismaClient.$metrics.json().then((metrics) => console.dir(metrics, { depth: Infinity })))
```

then i ran the job again. it failed on attempt 1, passed on attempt 2 (exponential backoff delay of 1 second). on the first attempt, 0 queries were successfully made so nothing logged (it timed out on the very first query attempt). on the 2nd attempt, all the queries were made. here are the metrics after the 1st query of the successful run:

first query:
```
{
  counters: [
    {
      key: 'prisma_client_queries_total',
      labels: {},
      value: 1,
      description: 'Total number of Prisma Client queries executed',
    },
    {
      key: 'prisma_datasource_queries_total',
      labels: {},
      value: 1,
      description: 'Total number of Datasource Queries executed',
    },
    {
      key: 'prisma_pool_connections_open',
      labels: {},
      value: 1,
      description: 'Number of currently open Pool Connections',
    },
  ],
  gauges: [
    {
      key: 'prisma_client_queries_active',
      labels: {},
      value: 0,
      description: 'Number of currently active Prisma Client queries',
    },
    {
      key: 'prisma_client_queries_wait',
      labels: {},
      value: 0,
      description: 'Number of queries currently waiting for a connection',
    },
    {
      key: 'prisma_pool_connections_busy',
      labels: {},
      value: 1,
      description: 'Number of currently busy Pool Connections (executing a database query)',
    },
    {
      key: 'prisma_pool_connections_idle',
      labels: {},
      value: 4,
      description:
        'Number of currently unused Pool Connections (waiting for the next pool query to run)',
    },
    {
      key: 'prisma_pool_connections_opened_total',
      labels: {},
      value: 1,
      description: 'Total number of Pool Connections opened',
    },
  ],
  histograms: [
    {
      key: 'prisma_client_queries_wait_histogram_ms',
      labels: {},
      value: {
        buckets: [[0, 0], [1, 2], [5, 0], [10, 0], [50, 0], [100, 0], [500, 0], [1000, 0], [5000, 0], [50000, 0]],
        sum: 0.002211,
        count: 2,
      },
      description: 'Histogram of the wait time of all queries in ms',
    },
    {
      key: 'prisma_datasource_queries_duration_histogram_ms',
      labels: {},
      value: {
        buckets: [[0, 0], [1, 0], [5, 0], [10, 0], [50, 1], [100, 0], [500, 0], [1000, 0], [5000, 0], [50000, 0]],
        sum: 12.715552,
        count: 1,
      },
      description: 'Histogram of the duration of all executed Datasource Queries in ms',
    },
  ],
}
```

as you can see, its the 1st connection pool used, nothing is waiting on the pool, and it all runs 1s after the initial failed run. the only place where 2 runs is indicated is in the `prisma_client_queries_wait_histogram_ms` where it says there were 2 runs - my guess is that this is a race condition since the 2nd query may have already started while the metrics logging was finishing

the query that was run is the following (same query that failed to run while waiting for the connection pool). in prisma its just a `prismaClient.action.findUniqueOrThrow({ where: { id: 'insert-our-uuid' } })`:
```
Query: SELECT "public"."Action"."id", "public"."Action"."createdAt", "public"."Action"."updatedAt", "public"."Action"."description", "public"."Action"."name", "public"."Action"."order", "public"."Action"."status", "public"."Action"."workflowId" FROM "public"."Action" WHERE ("public"."Action"."id" = $1 AND 1=1) LIMIT $2 OFFSET $3
Params: ["[insert-our-uuid"],1,0]
Duration: 12ms
```

**Comment by arnmishra** [comment]
upon a little more debugging, I found that another cloud run service makes a bunch of db calls at the same time. At the time of the example failure above, 28 db calls are made from that other service - each calls takes 1-7ms averaging around 2.5ms. none of them fail with any issues and the requests are never stuck waiting as far as I can tell (or if so its imperceptible, all these queries combined finish in <1s). 

I can't imagine how this would cause any of the failures above given this is using a different connection pool (also with 5 limit and 10s timeout) and nothing fails in that service (nor has it ever as far as I know). Just calling it out because its more information

**Comment by ludralph** [comment]
Hi @arnmishra 👋 

Thank you for taking the time to share all the details about your issue! This seems very strange indeed but we’re happy to look into this together with you 🙏.

Could you try to increase the database `connection_limit`? Also, what is the actual limit of database connections that’s configured on the database level? 

> Every time the query that fails is on the very first query made to the db in that container. At the time of the failure, there is no other activity in that container, and only a very minimal light load, if any, on the other 4 cloud run services.

We find this to be weird because we do not expect a connection pool timeout when there hasn't been any activity.

Have you considered increasing the [pool_timeout](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/connection-pool#setting-the-connection-pool-timeout) parameter? Can you also check Cloud SQL postgres database's performance metrics and limits to see if you are not hitting maximum connection limits?

**Comment by arnmishra** [comment]
@ludralph I'm happy to increase the timeout or pool size but I didn't want to blindly do that given it doesn't feel like we should be dealing with these timeouts. The operation it keeps timing out on is pretty time sensitive so just adding more delay felt like a dangerous solution for us in this case.

Looking at the cloud sql postgres stats, everything seems relatively standard (no issues with query latency, CPU, etc.). The connections to the db do peak at the time of the failure (from a usual 6-12 connections it peaks at 13 at the time of the failure). However, if I run `SELECT * FROM pg_settings WHERE name = 'max_connections'` I see the setting is a max of 400 connections to the db (and that is also documented here for an 8gb memory db https://cloud.google.com/sql/docs/postgres/flags#postgres-m). 

Honestly there is no way for it to get anywhere near that 400 number given even if we had all 5 cloud run services at max usage, the connection pools only allow 5 connections each so we'd peak at 25 connections. 

This screenshot shows the connections to the db, its the `dossStagingDb` (the teal line) and the time of failure was ~4:37PM. 
<img width="695" alt="Screenshot 2023-12-05 at 9 14 48 AM" src="https://github.com/prisma/prisma/assets/10519265/3cade433-a303-4f3a-89fa-95d902a4ec33">

**Reply by nikolasburk** [reply]
Hey @arnmishra  👋  thanks so much for bearing with us and sharing all the details about your app!

To recap, I believe you have:

- 5 instances of your app running on Google Cloud Run
- Each instance has a `connection_limit` of 5, which is the default value

If I understand correctly, this means that your DB is using at most 25 (out of 400) connections and Prisma Client fails when it can't open a 6th connection.

With 400 max connections and 5 app instances running, I think there should be quite a bit of room to increase the `connection_limit` parameter (e.g. to 50) to ensure Prisma Client's connection pool has access to more connections.

**Reply by arnmishra** [reply]
@nikolasburk close but not exactly. We do have 5 Prisma connection pools to connect to the db - 1 from each of the 5 cloud run instances. 

The issue isn't a 6th pool, the issue is when one of the cloud run instances attempts to use one of its connections (the first one). Ie this is the sec act state in the example above:

Cloud run 1 with pool 1: 5/5 connections active

Cloud run 2 with pool 2: 2/5 connections active

Cloud run 3 with pool 3: 0/5 connections active

Cloud run 4 with pool 4: 0/5 connections active

Cloud run 5 with pool 5: 0/5 connections active. A request comes to pool 5 so it attempts to make a connection to the db but times out despite minimal DB load.

**Reply by nikolasburk** [reply]
Thanks for your patience here, Arnav! We're escalating this to our Engineering team to get more input on the problem. In the meantime, could you try to set the `&pool_timeout=0` option on your connection string. This has solved the issue of another user in this [thread](https://github.com/prisma/prisma/issues/5691#issuecomment-780020860). To be clear: I wouldn't consider this a proper solution to your problem, merely a mitigation that might help prevent the app from failing until we have found the root cause for this issue.

**Reply by arnmishra** [reply]
@nikolasburk I think changing the pool timeout to 0 could be very risky in this case. The other thread you're pointing to has someone dealing with timeouts because their queries are super long - in our case the queries are nearly instantaneous so the issue isn't that the timeout isn't long enough. In the instances where the timeout isn't hit, the query completes successfully in <1s and in the times it fails, it takes the full 10s and never gets a connection. My hunch is if we change it to 0 we will end up in a process that just remains stuck

**Reply by arnmishra** [reply]
I'll see if I can give both bumped connection pool and 0s timeout a run today and see if i can reproduce with those but i think its unlikely those solve this issue

**Reply by nikolasburk** [reply]
> My hunch is if we change it to 0 we will end up in a process that just remains stuck

That's fair! 

> I'll see if I can give both bumped connection pool and 0s timeout a run today and see if i can reproduce with those but i think its unlikely those solve this issue

Thanks! We'll report back as soon as we hear something from Engineering!

**Reply by nikolasburk** [reply]
(FWIW, this is also a tricky for us because it's difficult/impossible to reproduce on our end. If _somehow_ you could provide a reproduction, that would of course be super useful.)

**Reply by arnmishra** [reply]
Totally understand that its going to be hard for you to reproduce given we don't know how to reproduce it ourselves either. 

Failures happens once every 5-10 runs or so. We have 3 retries to connect to the db configured. Usually on the failures it works on the 2nd try, the times it doesn't it usually fails on all 3 tries (so we have to do a 4th manual retry which so far has worked). Each try has a 1s pause between with exponential backoff. These requests happen at any time of the day and don't necessarily correlate with right after a deploy or anything else like that. All requests in all our other cloud run services are going through fine but there is no difference in the docker container that is running on cloud run itself. This is happening on both our staging and prod instances on the cloud run container for the same service (but not the same container itself - separate one on staging vs on production). In all these instances there is very limited load on the db and no stats spike on any of the metrics we see on cloud run or cloud sql. And the issue doesn't seem to be either a lack of time or a lack of available connections - instead it just seems like prisma isn't giving the process the available connection despite it being available. 

Not sure if there's someone on the prisma team I can jump on a call to pair with or something along those lines. I really just have no clue what else to try. I'll report back on whether we see anything working better with more connections or a 0s timeout.

**Reply by arnmishra** [reply]
To give the exact setup, this is what we have on our staging instance:
- 8gb memory postgres database running in cloud sql
- node server running in a docker container deployed on cloud run
- cloud run has 2 GiB memory with 1 vCPUs allocated to each instance of the container
- minimum of 1 instance is always available so no cold start (despite serverless env)
- prisma client is created during startup of the server using standard code:
```
const client = new PrismaClient({ 
  datasources: { 
    db: { 
      url: 'our url' 
    }
  }
})
```
- we have confirmed that only 1 client is created (log message above creation of the client - only logs once per deploy)
- the service uses BullMQ (https://docs.bullmq.io/) which essentially manages a queue running on redis
- the service starts up a bunch of "Bull workers" that read off the queue and execute actions based on events that get pulled off the queue
- when reading off the queue, the first thing we do is make a prisma `findUniqueOrThrow` call to access data from the db relating to the value we pulled off. It is during this first call that the query fails, if it ever does. It never fails on any subsequent query made.
- at the time of the failure, there are 5 connections to the pool, all available. There is no limitation on the database that is relevant and no significant load. The queries all take milliseconds. 
- As I understand it, despite all the above, the failure occurs in waiting for prisma's connection pool to give the query a connection to use as

**Reply by nikolasburk** [reply]
Thanks for following up with all that info! Quoting our PM for Prisma ORM here:

> This sounds like a network issue between container and database.
> The connection pool message can also mean that Prisma could not open a connection to the database in the background. As this is the first request to the pool, obviously the pool is empty and a connection has to be opened first - and if that goes crazy long, the pool gives up.
> I would go with the suggestion to increase the pool timeout indeed, but not infinity/0 but with 30 or 60s - maybe it is just the first connection from the container that is slow. So just letting the pool wait a bit longer might help. If the connection successfully opens, they can start figuring out why it is so slow.

I guess there's not too much new info in the though. Did you have any findings yesterday when you played around with the timeout and pool size options?

Another thing that would be interesting to understand for us is whether your DB actually receives a connection request from the connection pool. If that's not the case, we know that Prisma Client's query engine has trouble reaching the DB for some reason.

**Reply by arnmishra** [reply]
so this experiment has been even more confusing - I've run about 20-25 instances of it after adding timeout 0s and increasing the pool size to 10. none of them have failed and none of them have taken >10s (they are all finishing very quickly). 

I'll let it soak for another few days to see if we end up getting a failure but it was failing consistently and now isn't failing at all so somehow this seems to have fixed it? even though we aren't crossing the 10s timeout and we still are only using the 1st connection in the pool. 

After soak time, if we see no failure, ill go back to experimenting mid next week with either bumping pool size down to 0 or timeouts down to ~5s and see which causes failures. 

The network issue suggestion would be surprising to me because we have several cloud run containers in the same environment and none are having any network issues despite them all running effectively the same type of container. but i guess thats true with every other theory so something has to give haha.

Also it isn't the first connection in the container - we've had failures in the first connection after deploy but we've also had failures after the container has been active for a while (though no db calls were active at that time). unless maybe prisma refreshes the pool at some period of inactivity and its happening after the pool refreshes?

**Reply by janpio** [reply]
It might be that the database closes inactive/idle connections after some time - which indeed could put a connection pool in the original empty state again in some cases.

**Reply by arnmishra** [reply]
update: its still working on our staging environment as of today. im going to move the test to production with a 60s timeout instead of 0s timeout and see if it works from there. still very confusing why the timeout bump would fix this 🤔 but maybe im missing something in the infra

**Reply by nikolasburk** [reply]
Thanks a lot for the update! I agree that this still feels fishy but glad to hear that at least you're not seeing the failing instances any more 🙏 keep us posted about anything that seems weird with your app in that regard!

**Reply by arnmishra** [reply]
Just got it again `Timed out fetching a new connection from the connection pool. More info: http://pris.ly/d/connection-pool (Current connection pool timeout: 60, connection limit: 10)` - obviously seems more rare now but still doesn't make sense. I'll try bumping up to infinite again but i really don't have any confidence in this right now. im going to bump it back to infinite timeouts and might start handling timeouts outside of prisma if the infinite timeout works (wrap it with a 10s timeout of my own)

is there anything further to know about the internals of how the connection pool works? we've never been able to reproduce it ourselves but it feels anecdotally like most of the failures happen after a cloud run container has restarted (but not necessarily right after - it might've been 10+ hrs after restart, just has to be the first process run on the server). we're going to try adding an arbitrary prisma query after startup to see if we can get any consistency on the failure. however, we don't have conviction that its on startup anyways so not sure that'll help

**Reply by arnmishra** [reply]
the alternative we may consider is adding a very short timeout (ie 1 second) and then adding like 15-20 retries since every time it succeeds it does so very quickly

**Reply by arnmishra** [reply]
looks like this may be a cloud run infra issue related to cpu throttling per https://stackoverflow.com/questions/77694137/cloud-run-connection-to-postgres-intermittently-doesnt-work-on-first-attempt. going to see if we can solve via that instead, will report back (in case someone else is perusing this one day)

**Reply by RoopeHakulinen** [reply]
@arnmishra did disabling CPU throttling fix it for you?

**Reply by arnmishra** [reply]
yep that was it

**Reply by RoopeHakulinen** [reply]
Thanks for confirming. I'm experiencing this also with v2 Cloud Functions that use Cloud Run under the hood but don't seem to provide an option to set the CPU throttling.

**Reply by arnmishra** [reply]
just to check, did you see the stackoverflow suggestions? 
<img width="534" alt="Screenshot 2024-01-14 at 9 35 54 AM" src="https://github.com/prisma/prisma/assets/10519265/50d37c1f-b6ab-49da-9e11-348289404f33">
^that's the setting on cloud run that I used, not sure if cloud functions have the same settings accessible

**Reply by RoopeHakulinen** [reply]
Yeah, it is just that it is not available on Cloud Functions (at least via Terraform).. Thanks anyways!

**Reply by Kashuab** [reply]
Ran into the same problem, with the same exact setup. Disabling CPU throttling solved the problem for me. Thanks @arnmishra!

**Comment by arnmishra** [comment]
Is there any way to get eyes on this a little more urgently? Very risky situation for us and I can't imagine what is going wrong here and communicating once per day feels like it may take a while - happy to be available later in the day if that's helpful with timezones

**Reply by idbd1** [reply]
any news on that? I'm having a similar problem, tired a few things, but that didn't help.

**Reply by janpio** [reply]
Please open your own discussion @dominikabieder and describe your situation. Include as much information as possible, then we can take a look. Thanks. (This is not a general problem or bug, but depends 100% on the individual situation your app and database server is in.)

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
